### ISSUE TITLE
Also started on fpa core for beagle g

### ISSUE THREAD

**Issue by hstarmans** [lromor/beagleg-fpga-backend#8]
Hey all,

I am also working on FPGA core for beagle G.
The things i have working so far;
  - transactionalized fifo
  - divider, plan to use same algo as beagle g
  - spi interface
  - 
The code is written in nmigen and is quite a lot better then my earlier laser scanner code, which was more of a hot mess. You should check it out https://github.com/hstarmans/FPGAG/blob/master/src/FPGAG/core.py .
I am now adding movement, this has not been tested and is more of sketch but does compile.
Looking forward to your progress & comments.

Best,

Rik

**Comment by lromor** [comment]
Cool stuff! I think we it could be nice to integrate it and test it e2e in beagleg and verilator, what you think @hzeller ?

**Comment by hzeller** [comment]
In BeagleG, we have the branch 'hsg-experiments`: https://github.com/hzeller/beagleg/tree/hsg-experiments ('hardware step generation) in which there is a basic host-side SPI connection https://github.com/hzeller/beagleg/blob/hsg-experiments/src/spi-segment-queue.cc ... but right now it is essentially not doing SPI, but driving the verilator simulation behind, so this is why you see us wiggling the simulated clock. If you want to wire up your implementation, this is a good starting point.

Feature wise, we don't emit proper steps yet though, we got distracted with some other things we want to focus on (improving the planner).
Algorithm-wise here we were thinking of simplifying it compared to the BeagleG PRU implementation, which essentially calculates the time between steps and then waits. While if we have an FPGA, we can do something simpler as we're not tied to the limitations of the PRU: we can, in regluar intervals (say 100kHz) sample where each of the motors 'should' be and update a register - linar movement is just adding a number equivalent to speed/100kHz. Advantage is, that all operations (including acceleration and jerk) is just as simple as adding numbers (cheap operation, not many gates); we might need high precision, but also the advantage of FPGAs is, that doing arbitrary wide calculations is simple. Disadvantage is of course that we have to change more parts inside BeagleG while your implementation might be almost a drop-in replacement for the PRU.

We have a simple SPI protocol, in which the return data from the SPI tells us how full the queue is, so allows for some flow control. The first byte sent to SPI will return one byte in which it says how many free slots there are. So if this is zero, we have to try again later, but if there is space we can send; not super document right now, so code says more than words:
https://github.com/lromor/beagleg-fpga-backend/blob/master/host/beagleg-protocol.cc

Of course, longer term we'd like to also send back a signal outside the SPI bus to indicate that there is more space; that will then allow for more efficient sending from the computer without polling the bus.

Lenoardo and I are often video-hanging out when discussing these things. Might be interesting to have a session together and plan for some hacks we want to do together on BeagleG/FPGA @hstarmans ?

**Comment by hstarmans** [comment]
@hzeller yes it would be good to align so let me know when you guys meet up. I have used quite a lot of code from [luna]( https://github.com/greatscottgadgets/luna). 
I realise that a division might not be optimal use of the capabilities of the FPGA. A state register sounds nice. I know [haddington dynamics](https://www.hdrobotic.com/) uses an optical encoder and a FPGA to determine the position. So send me an invite at hstarmans gmail.com
